anndata/core/anndata.py

```diff
@@ -297,34 +269,38 @@ def __init__(
         if X is not None:
             self._X = X
             self._var = var
-            self._varm = varm
+            self._varm = AxisArrays(self, 1, varm)
         else:
             self._X = None if adata.isbacked else adata.X.copy()
             self._var = adata.var.copy()
-            self._varm = adata.varm.copy()
+            self._varm = AxisArrays(self, 1, adata.varm.copy())
 
     @property
     def X(self):
         if self._adata.isbacked:
-            if not self._adata.file.isopen: self._adata.file.open()
-            X = self._adata.file['raw.X']
-            if self._adata.isview: return X[self._adata._oidx, self._adata._vidx]
-            else: return X
-        else:
-            if self.n_obs == 1 and self.n_vars == 1:
-                warn_flatten()
-                return self._X[0, 0]
-            elif self.n_obs == 1 or self.n_vars == 1:
-                warn_flatten()
-                X = self._X
-                if issparse(self._X): X = self._X.toarray()
-                return X.flatten()
+            if not self._adata.file.isopen:
+                self._adata.file.open()
+            # Handle legacy file formats:
+            if "raw/X" in self._adata.file:
+                X = self._adata.file["raw/X"]
+            elif "raw.X" in self._adata.file:
+                X = self._adata.file['raw.X']  # Backwards compat
+            else:
+                raise AttributeError(
+                    f"Could not find dataset for raw X in file: {self._adata.file.filename}."
+                )
```

Comment:
I added this here since we needed a condition to throw an error if some accessed `adata.raw.X` if it didn't exist. `adata.X` should already throw an error the way it's accessed.

As to this existing in `_init_as_actual`, this was based on us checking the size of `raw` via it's `X`. This is different than `adata.X`. It's possible to read an `h5ad` even it doesn't have `X` defined, but should this be the case?